src/alda/AldaRequest.java

```diff
@@ -87,99 +86,71 @@ public String toJson() {
     return gson.toJson(this);
   }
 
-  /**
-   * Try up to WRONG_JOB_RETRIES times to receive a message from the server that
-   * has the same jobId as the request we sent.
-   *
-   * This is to avoid accidentally using a response that was sent out of sync
-   * for a different request.
-   *
-   * Returns null if we fail to receive a message for our jobId.
-   */
-  private AldaResponse receiveMatchingResponse(Socket client, int timeout)
+  private AldaResponse sendRequest(ZMsg request, int timeout, int retries)
     throws NoResponseException {
+    if (retries < 0 || Thread.currentThread().isInterrupted()) {
+      request.destroy();
+      throw new NoResponseException("Alda server is down. To start the server, run `alda up`.");
+    }
+
     Poller poller = getPoller();
+    Socket client = rebuildSocket();
 
+    // When non-null, used to help ensure that we don't use a response from the
+    // server that was for a different request.
     String jobId = options == null ? null : options.jobId;
 
-    int retries = WRONG_JOB_RETRIES;
+    // don't destroy the message after sending
+    request.send(client, false);
 
-    while (retries > 0) {
-      int rc = poller.poll(timeout);
-      if (rc == -1) {
-        throw new NoResponseException("Connection interrupted.");
-      }
-
-      if (poller.pollin(0)) {
-        ZMsg zmsg = ZMsg.recvMsg(client);
-        if (debug) zmsg.dump();
+    int rc = poller.poll(timeout);
+    if (rc == -1) {
+      request.destroy();
+      throw new NoResponseException("Connection interrupted.");
+    }
 
-        byte[] address = zmsg.unwrap().getData(); // discard envelope
-        String responseJson = zmsg.popString();
+    if (poller.pollin(0)) {
+      ZMsg zmsg = ZMsg.recvMsg(client);
+      if (debug) zmsg.dump();
 
-        AldaResponse response = AldaResponse.fromJson(responseJson);
+      byte[] address = zmsg.unwrap().getData(); // discard envelope
+      String responseJson = zmsg.popString();
 
-        if (jobId != null &&
-            response.jobId != null &&
-            !response.jobId.equals(jobId)) {
-          retries--;
-          continue;
-        }
+      AldaResponse response = AldaResponse.fromJson(responseJson);
 
+      // If there is a jobId option, we will ignore any response from the server
+      // that doesn't have the same jobId, and try again. This will not count
+      // against our remaining retries, as the server did respond.
+      if (jobId == null ||
+          (response.jobId != null && response.jobId.equals(jobId))) {
         if (!response.noWorker)
           response.workerAddress = zmsg.pop().getData();
 
+        request.destroy();
         return response;
       }
 
-      retries--;
+      // Wrong jobId; try again without decrementing retries.
+      return sendRequest(request, timeout, retries);
```

Comment:
I'm able to get this function to stackoverflow if I start the alda server and run a `piano: c d e` immediately (assuming that the server takes a while to come up). 

When I'm getting this error `response.jobId` is null, so maybe we can decrement the retries only when this is null?